replay/src/replay.ts

```diff
@@ -34,17 +85,159 @@ export class Replay extends Subprocess implements IReplayIpcNodeDelegate {
+         start.setTime(this.startTime!.getTime());
+      }
+      this.sendMetricsToS3(start, end);
+
+      if (finished) {
+        this.stop(false); // just once for now
+      }
    }
 
    protected async teardown(): Promise<void> {
+      logger.info(`Replay ${this.id}: teardown`);
+
+      if (this.error === false) {
+         await Replay.updateReplayStatus(this.id, ChildProgramStatus.DONE);
+      }
+
       this.ipcNode.stop();
    }
 
+   private async getWorkload() {
+
+      logger.info(`Getting workload from storage`);
+      this.workloadLocation = await Replay.getWorkloadForCapture(this.config.captureId);
+      this.workload = await this.storage.readJson(this.workloadLocation) as any;
+   }
+
+   // queryInInterval takes the index of the query in the workload and the time
+   //                 that next loop will begin in milliseconds and returns true
+   //                 if the query should be scheduled for the current loop otherwise false.
+   private queryInInterval(index: number, intervalStartTime: number): boolean {
+
+      const delay = this.getDelayForIndex(index, intervalStartTime);
+      return (delay >= 0 && delay < this.interval) ? true : false  ;
+   }
+
+   // getDelayForIndex takes in an index and the current interval's start time and returns the
+   //                  number of milliseconds to delay from the interval's start time.
+   private getDelayForIndex(index: number, intervalStartTime: number): number {
+      if (this.workload == null) {
+         return 0;
+      } else if (index > 0 && index < this.workload!.length) {
+         const queryStartTime: Date = new Date(this.workload![index].event_time);
+         const workloadStartTime: Date = new Date(this.workload![0].event_time);
+         const relativeQueryST = queryStartTime.getTime() - workloadStartTime.getTime() + this.startTime!.getTime();
+         const delay = relativeQueryST - intervalStartTime + (Date.now() - intervalStartTime);
+
+         return delay;
+      } else {
+         return 0;
+      }
+   }
+
+   private processQuery(type: any, argument: any) {
+
+      if (type === "Query") {
+
+        const remoteReplayDbConfig = require('../db/remoteReplayConfig.json');
+        const conn = mysql.createConnection(remoteReplayDbConfig);
+
+        return new Promise<any>((resolve, reject) => {
+          conn.connect((connErr) => {
+              if (connErr) {
+                reject(connErr);
+              } else {
+                const updateStr = mysql.format(argument, []);
+                conn.query(updateStr, (updateErr, rows) => {
+                    conn.end();
+                    if (updateErr) {
+                      reject(updateErr);
+                    } else {
+                      resolve(rows);
+                    }
+                });
+              }
+          });
+        });
+
+      } else { return null; }
+   }
+   private async sendMetricsToS3(start: Date, end: Date, firstTry: boolean = true) {
```

Comment:
looks pretty similar to the capture's sendMetricsToS3 function, we might want to consider refactoring to prevent duplicate code